mirrord/tls-util/src/san.rs

```diff
@@ -0,0 +1,74 @@
+use std::net::IpAddr;
+
+use rustls::pki_types::{CertificateDer, ServerName};
+use tracing::Level;
+use x509_parser::{
+    pem::Pem,
+    prelude::{FromDer, GeneralName, X509Certificate},
+};
+
+use crate::error::GetSanError;
+
+/// Convenience trait for extracing [`ServerName`]s from an x509 certificate.
+pub trait HasSubjectAlternateNames {
+    fn subject_alternate_names(&self) -> Result<Vec<ServerName<'static>>, GetSanError>;
+}
+
+impl HasSubjectAlternateNames for CertificateDer<'_> {
+    #[tracing::instrument(level = Level::DEBUG, skip(self), ret, err(level = Level::DEBUG))]
+    fn subject_alternate_names(&self) -> Result<Vec<ServerName<'static>>, GetSanError> {
+        let certificate = X509Certificate::from_der(self)?.1;
+        get_san(&certificate)
+    }
+}
+
+impl HasSubjectAlternateNames for Pem {
+    #[tracing::instrument(level = Level::DEBUG, skip(self), ret, err(level = Level::DEBUG))]
+    fn subject_alternate_names(&self) -> Result<Vec<ServerName<'static>>, GetSanError> {
+        let certificate = self.parse_x509()?;
+        get_san(&certificate)
+    }
+}
+
+fn get_san(certificate: &X509Certificate<'_>) -> Result<Vec<ServerName<'static>>, GetSanError> {
```

Comment:
Not very important, but why not `impl HasSubjectAlternateNames for X509Certificate<'_>`?